src/flow/src/compute/render/reduce.rs

```diff
@@ -0,0 +1,957 @@
+            } else {
+                // ignore for whatever reason
+                continue;
+            }
+        };
+
+        let mut accum_output = AccumOutput::new();
+        eval_simple_aggrs(
+            simple_aggrs,
+            &accums,
+            &accum_ranges,
+            &col_diffs,
+            &mut accum_output,
+            err_collector,
+        );
+
+        // for distinct input
+        eval_distinct_aggrs(
+            distinct_aggrs,
+            distinct_input,
+            &accums,
+            &accum_ranges,
+            &col_diffs,
+            &mut accum_output,
+            SubgraphArg {
+                now,
+                err_collector,
+                scheduler,
+                send,
+            },
+        );
+
+        // get and append results
+        err_collector.run(|| {
+            let (new_accums, res_val_row) = accum_output.into_accum_output()?;
+
+            // construct the updates and save it
+            all_updates.push(((key.clone(), Row::new(new_accums)), now, 1));
+            let mut key_val = key;
+            key_val.extend(res_val_row);
+            all_outputs.push((key_val, now, 1));
+            Ok(())
+        });
+    }
+    err_collector.run(|| {
+        arrange.apply_updates(now, all_updates)?;
+        arrange.compaction_to(now)
+    });
+
+    // for all arranges involved, schedule next time this subgraph should run
+    // no future updates should exist here
+    let all_arrange_used = distinct_input
+        .iter()
+        .flatten()
+        .map(|d| d.write())
+        .chain(std::iter::once(arrange));
+    check_no_future_updates(all_arrange_used, err_collector, now);
+
+    send.give(all_outputs);
+}
+
+/// Eval simple aggregate functions with no distinct input
+fn eval_simple_aggrs(
+    simple_aggrs: &Vec<AggrWithIndex>,
+    accums: &[Value],
+    accum_ranges: &[Range<usize>],
+    col_diffs: &[Vec<(Value, i64)>],
+    accum_output: &mut AccumOutput,
+    err_collector: &ErrCollector,
+) {
+    for AggrWithIndex {
+        expr,
+        input_idx,
+        output_idx,
+    } in simple_aggrs
+    {
+        let cur_accum_range = accum_ranges[*output_idx].clone(); // range of current accum
+        let cur_old_accum = accums
+            .get(cur_accum_range)
+            .unwrap_or_default()
+            .iter()
+            .cloned();
+        let cur_col_diff = col_diffs[*input_idx].iter().cloned();
+
+        // actual eval aggregation function
+        if let Some((res, new_accum)) =
+            err_collector.run(|| expr.func.eval_diff_accumulable(cur_old_accum, cur_col_diff))
+        {
+            accum_output.insert_accum(*output_idx, new_accum);
+            accum_output.insert_output(*output_idx, res);
+        } // else just collect error and continue
+    }
+}
+
+#[derive(Debug)]
+struct AccumOutput {
+    accum: BTreeMap<usize, Vec<Value>>,
+    output: BTreeMap<usize, Value>,
+}
+
+impl AccumOutput {
+    fn new() -> Self {
```

Comment:
Looks like missing a construct such as `with_capacity(uszie)` when we known the exact capacity.

Reply:
> Looks like missing a construct such as `with_capacity(uszie)` when we known the exact capacity.

Since BTreeMap lack a `with_capacity` method(see https://users.rust-lang.org/t/btreemap-with-capacity/39770/3) I'm not sure it would be necessary, although  maybe I should use `Vec<Option<>>` instead...